File: app/src/main/java/com/dmi/perfectreader/bookcontrol/BookControlFragment.java
package com.dmi.perfectreader.bookcontrol;

import android.support.annotation.NonNull;
import android.view.KeyEvent;
import android.view.MotionEvent;
import android.view.View;
import android.widget.FrameLayout;

import com.dmi.perfectreader.R;
import com.dmi.perfectreader.book.BookPresenter;
import com.dmi.perfectreader.bookreader.BookReaderFragment;
import com.dmi.util.base.BaseFragment;
import com.dmi.util.layout.HasLayout;

import javax.inject.Inject;

import butterknife.Bind;
import dagger.ObjectGraph;
import dagger.Provides;

@HasLayout(R.layout.fragment_bookcontrol)
public class BookControlFragment extends BaseFragment implements View.OnTouchListener {
    @Bind(R.id.spaceView)
    protected FrameLayout spaceView;

    @Inject
    protected BookControlPresenter presenter;

    private final TouchInfo touchInfo = new TouchInfo();

    @Override
    protected ObjectGraph createObjectGraph(ObjectGraph parentGraph) {
        return parentGraph.plus(new Module());
    }

    @Override
    public BookControlPresenter presenter() {
        return presenter;
    }

    @Override
    public void onViewCreated(@NonNull View view) {
        super.onViewCreated(view);
        spaceView.setOnTouchListener(this);
    }

    @Override
    public boolean onTouch(View view, MotionEvent event) {
        touchInfo.width = spaceView.getWidth();
        touchInfo.height = spaceView.getHeight();
        touchInfo.x = event.getX();
        touchInfo.y = event.getY();
        touchInfo.touchDiameter = event.getTouchMajor();

        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                presenter.onTouchDown(touchInfo);
                break;
            case MotionEvent.ACTION_MOVE:
                presenter.onTouchMove(touchInfo);
                break;
            case MotionEvent.ACTION_UP:
                presenter.onTouchUp(touchInfo);
                break;
        }

        return true;
    }

    @Override
    public boolean onKeyDown(int keyCode, @NonNull KeyEvent event) {
        presenter().onKeyDown(HardKey.fromKeyCode(keyCode));
        return true;
    }

    @Override
    public boolean onKeyUp(int keyCode, @NonNull KeyEvent event) {
        return true;
    }

    @dagger.Module(addsTo = BookReaderFragment.Module.class, injects = {
            BookControlFragment.class,
            BookControlPresenter.class,
    })
    public class Module {
        @Provides
        public BookControlFragment view() {
            return BookControlFragment.this;
        }

        @Provides
        public BookPresenter bookPresenter() {
            BookReaderFragment bookReaderFragment = parentFragment();
            return bookReaderFragment.book().presenter();
        }
    }
}

File: app/src/main/kotlin/com/dmi/perfectreader/bookcontrol/BookControlFragment.kt
package com.dmi.perfectreader.bookcontrol

import android.view.KeyEvent
import android.view.MotionEvent
import android.view.View
import android.widget.FrameLayout
import butterknife.Bind
import com.dmi.perfectreader.R
import com.dmi.perfectreader.book.BookPresenter
import com.dmi.perfectreader.bookreader.BookReaderFragment
import com.dmi.util.base.BaseFragment
import com.dmi.util.layout.HasLayout
import dagger.ObjectGraph
import dagger.Provides
import javax.inject.Inject

@HasLayout(R.layout.fragment_bookcontrol)
open class BookControlFragment : BaseFragment(), View.OnTouchListener {
    @Bind(R.id.spaceView)
    protected lateinit var spaceView: FrameLayout

    @Inject
    protected lateinit var presenter: BookControlPresenter

    private val touchInfo = TouchInfo()

    override fun createObjectGraph(parentGraph: ObjectGraph): ObjectGraph {
        return parentGraph.plus(Module())
    }

    public override fun presenter(): BookControlPresenter {
        return presenter
    }

    override fun onViewCreated(view: View) {
        super.onViewCreated(view)
        spaceView.setOnTouchListener(this)
    }

    override fun onTouch(view: View, event: MotionEvent): Boolean {
        touchInfo.width = spaceView.width.toFloat()
        touchInfo.height = spaceView.height.toFloat()
        touchInfo.x = event.x
        touchInfo.y = event.y
        touchInfo.touchDiameter = event.touchMajor

        when (event.action) {
            MotionEvent.ACTION_DOWN -> presenter.onTouchDown(touchInfo)
            MotionEvent.ACTION_MOVE -> presenter.onTouchMove(touchInfo)
            MotionEvent.ACTION_UP -> presenter.onTouchUp(touchInfo)
        }

        return true
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        presenter().onKeyDown(HardKey.fromKeyCode(keyCode))
        return true
    }

    override fun onKeyUp(keyCode: Int, event: KeyEvent): Boolean {
        return true
    }

    @dagger.Module(addsTo = BookReaderFragment.Module::class, injects = arrayOf(BookControlFragment::class, BookControlPresenter::class))
    inner class Module {
        @Provides
        fun view(): BookControlFragment {
            return this@BookControlFragment
        }

        @Provides
        fun bookPresenter(): BookPresenter {
            val bookReaderFragment = parentFragment<BookReaderFragment>()
            return bookReaderFragment.book().presenter()
        }
    }
}
